Dealing with missing contests and selections

shreyasminocha opened the pull request:
I first noticed this a couple weeks ago when I was working on expanding our test suite. The 'sync-async identical' test fails when ballots with missing contests come into the picture. The sync version auto-adds a contest full of placeholders, but the async version gets the contests from the ballots instead of getting them from the manifest, so it doesn't do such a thing. We could adjust either to do the other. @danwallach I'd appreciate your input.

Also, to get the 'encryption/decryption inverses' test to pass, I had to slightly adjust the definition of plaintext ballot/contest equality. It raises interesting questions about whether implicitly false, explicitly false, and placeholder false selections are equivalent; I went with "they're all equivalent".

Fixes #39

danwallach commented:
I'll investigate later. Indeed, we want the `async` code to behave identically to the regular version.

danwallach commented:
Discussion summary:

- A missing contest in a `PlaintextBallot` is an error and should throw an exception.
- A missing contest in the async-encryption version should be treated as an undervote. So the corresponding `PlaintextContest` will need to be created and encrypted. This will slow down the process, but we can't avoid it.
- The unit tests should assert equality between synchronous and async encryption, without any filtering.

danwallach commented:
Just had a conversation with the EnhancedVoting people. It's okay for us to throw an error if we have a missing contest. They have a linear UI that drives the voter past each contest in sequence, so they'll definitely be passing us contests each time.

So, that means we DO need to check for missing contests (and, while we're at it, for contests outside the ballot style), but we DON'T need to do any last minute encryptions of undervoted contests. These conditions only need to throw an error.

danwallach commented:
Hmm. I think we should define a PlaintextContest with missing selections as an error and throw an exception.

shreyasminocha commented:
> define a PlaintextContest with missing selections as an error and throw an exception

That seems to break the -python compatibility tests. The python generators are set up to generate contests that may have missing selections.

danwallach commented:
Uggh. Does Python's equality check do this kind of missing selection logic?

My gut reaction is that I want our library to be more strict, when there's ambiguity, and get the correct answer when it matters.

shreyasminocha commented:
> Does Python's equality check do this kind of missing selection logic?

The `__eq__` methods don't, so we probably shouldn't either; AFAIK though, those aren't used in their round-trip tests. My understanding is that they assert that for every selection in the original ballot, the `vote` field of the corresponding decrypted selection matches (but they don't check the converse).

I'm usually a fan of strictness, but IMO here it's also important that we can process the sample data. A missing selection being an implicit `0` seems reasonable. How about we adjust the failing test instead?

danwallach commented:
Ok. We'll accept the slightly malformed input. What we should do, though, is have a "normalize" function that runs first, making a copy that's valid, and then pass that along. We could then have tests that say "weirdBallot.normalize().equals(goodBallot())"

shreyasminocha commented:
All done. You probably want to review this one when you get a chance.

shreyasminocha commented:
- Moved the normalization stuff to non-exported functions
- they now loop over the manifest's fields
- they throw errors if there are extraneous contests/selections
- Added return type annotations and doc comments.

I don't think we can call normalize in the `equals` methods because normalization requires manifest-related parameters, which, as we discussed, aren't currently associated with the `Plaintext*` classes.

danwallach commented:
Ahh, good point about the Manifest stuff. I'm not sure this is the sort of thing that we want to make visible to external API users. It's good enough that we have it for ourselves, at least for now.
